src/tenant_tokens.rs

```diff
@@ -0,0 +1,128 @@
+use crate::{
+    errors::* 
+};
+use serde::{Serialize, Deserialize};
+use jsonwebtoken::{encode, Header, EncodingKey};
+use time::{OffsetDateTime};
+use serde_json::Value;
+
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")] 
+struct TenantTokenClaim {
+    api_key_prefix: String,
+    search_rules: Value,
+    #[serde(with = "time::serde::timestamp::option")]
+    exp: Option<OffsetDateTime>,
+}
+
+pub fn generate_tenant_token(search_rules: Value, api_key: impl AsRef<str>, expires_at: Option<OffsetDateTime>) -> Result<String, Error> {
+    if api_key.as_ref().chars().count() < 8 {
+        return Err(Error::TenantTokensInvalidApiKey)
+    }
+
+    if expires_at.map_or(false, |expires_at| OffsetDateTime::now_utc() > expires_at) {
+        return Err(Error::TenantTokensExpiredSignature)
+    }
+
+    let key_prefix = api_key.as_ref().chars().take(8).collect();
+    let claims = TenantTokenClaim {
+        api_key_prefix: key_prefix,
+        exp: expires_at,
+        search_rules
+    };
+
+    let token = encode(
+        &Header::default(),
+        &claims,
+        &EncodingKey::from_secret(api_key.as_ref().as_bytes()),
+    );
+
+    Ok(token?)
+}
+
+#[cfg(test)]
+mod tests {
+    use serde_json::json;
+    use crate::tenant_tokens::*;
+    use jsonwebtoken::{decode, DecodingKey, Validation, Algorithm};
+    use meilisearch_test_macro::meilisearch_test;
```

Comment:
```suggestion
```

All the tests in this file do not concern meilisearch and are not async. Thus can create normal tests.